route: Support refer route next hop interface by profile name

cathay4t opened the pull request:
Example YAML:

```yaml
---
interfaces:
  - name: port1
    type: ethernet
    mac-address: 00:23:45:67:89:1a
    identifier: mac-address
    ipv4:
      enabled: true
      dhcp: false
      address:
        - ip: 192.0.2.251
          prefix-length: 24
routes:
  config:
  - destination: 0.0.0.0/0
    next-hop-address: 192.0.2.1
    next-hop-interface: port1
```

For route next hop interface, we will prefer kernel name but fallback to
search profile name. Will raise error when route is referring to
multiple interface sharing the same profile name.

Integration test cases included.

Resolves: https://issues.redhat.com/browse/RHEL-32495
Resolves: https://issues.redhat.com/browse/RHEL-56558

liangwen12year commented on rust/src/lib/route.rs:
```diff
@@ -104,6 +105,51 @@ impl Routes {
+        if let Some(config_routes) = self.config.as_mut() {
+            for route in config_routes.iter_mut().filter(|r| !r.is_absent()) {
+                let route_display_str = route.to_string();
+                if let Some(nex_hop_iface) = route.next_hop_iface.as_mut() {
```
How about changing the variable to `next_hop_iface` to improve clarity?

cathay4t replied:
Fixed.

liangwen12year commented on rust/src/lib/route.rs:
```diff
@@ -104,6 +105,51 @@ impl Routes {
+                        return Err(NmstateError::new(
+                            ErrorKind::InvalidArgument,
+                            format!(
+                                "Route '{}': next hop interface {} no found",
```
How about changing it to "Route '{}': next hop interface {} not found"?

cathay4t replied:
I don't see a difference between your suggested string and my code.

ihuguet replied:
~no~ not

liangwen12year commented on rust/src/lib/route.rs:
```diff
@@ -104,6 +105,51 @@ impl Routes {
+            for route in config_routes.iter_mut().filter(|r| !r.is_absent()) {
+                let route_display_str = route.to_string();
+                if let Some(nex_hop_iface) = route.next_hop_iface.as_mut() {
+                    let kernel_names = iface_name_search.get(nex_hop_iface);
```
As a potential optimization, maybe cache the results of `iface_name_search.get(nex_hop_iface)` if the same `nex_hop_iface` is used multiple times?

cathay4t replied:
No cache required. The `iface_name_search` is a HashMap which is O(1) complexity for `get()` action.

liangwen12year commented on rust/src/lib/route.rs:
```diff
@@ -104,6 +105,51 @@ impl Routes {
+                            format!(
+                                "Route '{}' defined with next \
+                                hop interface {} but \
+                                multiple interfaces are sharing the \
```
`the` is redundant here?

cathay4t replied:
Fixed. Thanks.

ihuguet commented on rust/src/lib/route.rs:
```diff
@@ -104,6 +105,51 @@ impl Routes {
+        let iface_name_search = InterfaceNameSearch::new(merged_ifaces);
+
+        if let Some(config_routes) = self.config.as_mut() {
+            for route in config_routes.iter_mut().filter(|r| !r.is_absent()) {
```
Why the `filter(|r| !r.is_absent())`? To mark as absent it is also OK to use the profile name, IMO.

cathay4t replied:
You are correct. Fixed.

ihuguet commented on tests/integration/route_test.py:
Please add test(s) for absent route

cathay4t replied:
Done.

ihuguet commented on rust/src/lib/route.rs:
```diff
@@ -104,6 +105,51 @@ impl Routes {
+
+        if let Some(config_routes) = self.config.as_mut() {
+            for route in config_routes.iter_mut().filter(|r| !r.is_absent()) {
+                let route_display_str = route.to_string();
+                if let Some(next_hop_iface) = route.next_hop_iface.as_mut() {
```
I'd like to avoid allocating the string if it's not needed. I see that it's because you take its next_hop_iface as mut ref.

Maybe you can do something like: 
```
let new_next_hop_iface = if let Some(next_hop_iface) = route.next_hop_iface.as_ref() {
    ...
                    } else if let Some(kernel_name) = kernel_names.first() {
                        Some(kernel_name.to_string())
                    } else {
    ...
} else {
    None
}

if let Some(new_next_hop_iface) = new_next_hop_iface {
    *route.next_hop_iface.as_mut() = new_next_hop_iface;
}
```

cathay4t replied:
You cannot has `as_ref()` inside of
```rust
        if let Some(config_routes) = self.config.as_mut() {
            for route in config_routes.iter_mut() {
```

cathay4t replied:
Let me find a way to only use `to_string()` when error happens.

cathay4t replied:
Changed to
```rust
        if let Some(config_routes) = self.config.as_mut() {
            for route in config_routes.iter_mut() {
                if let Some(next_hop_iface) = route.next_hop_iface.as_mut() {
                    let kernel_names = iface_name_search.get(next_hop_iface);
                    // Prefer kernel name as port name
                    if kernel_names.contains(&next_hop_iface.as_str()) {
                        continue;
                    } else if let Some(kernel_name) = kernel_names.first() {
                        *next_hop_iface = kernel_name.to_string();
                        continue;
                    }
                }
                if let Some(next_hop_iface) = route.next_hop_iface.as_ref() {
                    let kernel_names = iface_name_search.get(next_hop_iface);
                    if kernel_names.len() > 1 {
                        return Err(NmstateError::new(
                            ErrorKind::InvalidArgument,
                            format!(
                                "Route '{}' defined with next \
                                hop interface {} but \
                                multiple interfaces are sharing \
                                this profile name",
                                route,
                                next_hop_iface.as_str()
                            ),
                        ));
                    } else if kernel_names.is_empty() {
                        return Err(NmstateError::new(
                            ErrorKind::InvalidArgument,
                            format!(
                                "Route '{}': next hop interface {} not found",
                                route,
                                next_hop_iface.as_str()
                            ),
                        ));
                    }
                }
            }
        }

```

No extra memory allocation except generating error message.

ihuguet replied:
This compiles for me. I'm not sure if the logic is 100% correct, but the borrow checker is happy with my approach.
```diff
diff --git a/rust/src/lib/route.rs b/rust/src/lib/route.rs
index 95b60210..ef8add99 100644
--- a/rust/src/lib/route.rs
+++ b/rust/src/lib/route.rs
@@ -114,31 +114,18 @@ impl Routes {
 
         if let Some(config_routes) = self.config.as_mut() {
             for route in config_routes.iter_mut() {
-                if let Some(next_hop_iface) = route.next_hop_iface.as_mut() {
+                let new_iface_name = if let Some(next_hop_iface) =
+                    route.next_hop_iface.as_ref()
+                {
                     let kernel_names = iface_name_search.get(next_hop_iface);
                     // Prefer kernel name as port name
                     if kernel_names.contains(&next_hop_iface.as_str()) {
                         continue;
-                    } else if let Some(kernel_name) = kernel_names.first() {
-                        if kernel_names.len() == 1 {
-                            *next_hop_iface = kernel_name.to_string();
-                            continue;
-                        }
-                    }
-                    // We do not raise error because we cannot display the
-                    // route when holding route.next_hop_iface.as_mut().
-                    // The error will be generated by following if block.
-                }
-                if let Some(next_hop_iface) = route.next_hop_iface.as_ref() {
-                    let kernel_names = iface_name_search.get(next_hop_iface);
-                    if kernel_names.len() > 1 {
+                    } else if kernel_names.len() > 1 {
                         return Err(NmstateError::new(
                             ErrorKind::InvalidArgument,
                             format!(
-                                "Route '{}' defined with next \
-                                hop interface {} but \
-                                multiple interfaces are sharing \
-                                this profile name",
+                                "Route '{}': multiple interfaces sharing the {} profile name",
                                 route,
                                 next_hop_iface.as_str()
                             ),
@@ -152,7 +139,17 @@ impl Routes {
                                 next_hop_iface.as_str()
                             ),
                         ));
+                    } else if let Some(kernel_name) = kernel_names.first() {
+                        Some(kernel_name.to_string())
+                    } else {
+                        None
                     }
+                } else {
+                    None
+                };
+
+                if let Some(new_iface_name) = new_iface_name {
+                    route.next_hop_iface.replace(new_iface_name);
                 }
             }
         }
```

cathay4t replied:
Got it. Your patch is better.

cathay4t commented:
The failure of nm_stable on slow group is known issue of NetworkManager-1.51.90-2.el9 which should be fixed by pending update of NM.

cathay4t committed:
route: Support refer route next hop interface by profile name

Example YAML:

```yaml
---
interfaces:
  - name: port1
    type: ethernet
    mac-address: 00:23:45:67:89:1a
    identifier: mac-address
    ipv4:
      enabled: true
      dhcp: false
      address:
        - ip: 192.0.2.251
          prefix-length: 24
routes:
  config:
  - destination: 0.0.0.0/0
    next-hop-address: 192.0.2.1
    next-hop-interface: port1
```

For route next hop interface, we will prefer kernel name but fallback to
search profile name. Will raise error when route is referring to
multiple interface sharing the same profile name.

Integration test cases included.

Resolves: https://issues.redhat.com/browse/RHEL-32495
Resolves: https://issues.redhat.com/browse/RHEL-56558

Signed-off-by: Gris Ge <fge@redhat.com>
